rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,241 @@
+This allows locally extending a shared store that is periodically extended with additional store objects.
+
+# Motivation
+[motivation]: #motivation
+
+TODO: link Replit blog post.
+
+## Technical motivation
+
+Many organizational users of Nix have a large collection of Nix store objects they wish to share with a number of consumers, be they human users, build farm workers etc.
+The existing ways of doing this are:
+
+- Share nothing: Copy store objects to each consumer
+
+- Share everything: Use single mutable NFS store, carefully synchronizing updates.
+
+- Overlay everything: Mount all of `/nix` (store and DB) with OverlayFS.
+
+Each has serious drawbacks:
+
+- Share nothing wastes tones of space as many duplicate store objects are stored separately.
+
+- Share everything incurs major overhead from synchronization, even if consumers are making store objects they don't intend any other consumer to use.
+
+- Overlay everything cannot take advantage of new store objects added to the lower store, because its "fork" of the DB covers up the lower store's.
+
+The new `local-overlay` store also uses OverlayFS, but just for the store directory.
+The database is still a regular fresh empty one to start, and instead Nix explicitly knows about the lower store, so it can get any information for the DB it needs from it manually.
+This avoids all 3 downsides:
+
+- Store objects are never deduplicated.
+  OverlayFS ensures that one copy in either store dir layer is enough, and we are careful never to wastefully include a store object in both layers.
+
+- No excess synchronization.
+  local changes are just that: local, not shared with any other consumer.
+  The lower store is never written to (no modifications or even filesystem locks) so any slow NFS write and sync paths should not be encountered.
+
+- No rigidity of the lower store.
+  Since Nix sees both the `overlay-store`'s DB and the lower store, it is possible for it to be aware of and utilize objects that were added to the lower store after the upper store was created.
+
+This gives us a "best of all three worlds" solution.
+
+## Marketing motivation
+
+It is quite common for organizations using Nix to first adopt it behind the scenes.
+That is to say, Nix is used to prepare some artifacts which are then presented to a consumer that need not be aware they were made with Nix.
+Later though, because of Nix's gaining popularity, there may be a desire to reveal its usage so consumers can use Nix themselves.
+Rather than Nix being a controversial tool worth hiding, it can be a popular tool worth exposing.
+Nix-unware usage can still work, but Nix-aware usage can do additional things.
+
+The `local-overlay` store can serve as a crucial tool to bridge these two modes of using Nix.
+The lower store can be however the artifacts were disseminated in the "hidden Nix" first phase of adoption, perhaps with a small tweak to expose the DB / daemon socket if it wasn't before.
```

Comment:
Could the first part of the sentence be reworked a little? Initially I was really tripped up by trying to read the 'however' as an isolated word, but I see now it's part of a bigger clause that follows:

> The lower store can be _however-the-artifacts-were-disseminated_ in the "hidden Nix" first phase of adoption